source/2_platform/rex_windows/src/input/input.cpp

```diff
@@ -0,0 +1,72 @@
+#include "rex_windows/input/input.h"
+
+#include "rex_engine/diagnostics/win/win_call.h"
+#include "rex_std/algorithm.h"
+#include "rex_std/assert.h"
+#include "rex_std/vector.h"
+
+#define NOMINMAX
+#include <Windows.h>
+
+namespace rex
+{
+  namespace input
+  {
+    // NOLINTBEGIN(cppcoreguidelines-avoid-non-const-global-variables)
+    card32 g_current_input_idx               = 1;
+    rsl::array<bool, 256> g_keys_even_frames = {};
+    rsl::array<bool, 256> g_keys_odd_frames  = {};
+    rsl::array<bool, 256>* g_prev_keys       = &g_keys_even_frames;
+    rsl::array<bool, 256>* g_current_keys    = &g_keys_odd_frames;
+    // NOLINTEND(cppcoreguidelines-avoid-non-const-global-variables)
+
+    uint8 poll()
+    {
+      constexpr card32 max_input_keys = (rsl::numeric_limits<uint8>::max)();
+      for(card32 i = g_current_input_idx; i < max_input_keys; ++i, ++g_current_input_idx)
+      {
+        uint8 key = static_cast<uint8>(i);
+
+        if(WIN_CALL(GetAsyncKeyState(key)) & 0x8000u) // NOLINT(hicpp-signed-bitwise)
+        {
+          ++g_current_input_idx;
+          return key;
+        }
+      }
+
+      g_current_input_idx = 1;
+      return '\0';
+    }
+
+    void update()
+    {
+      rsl::swap(g_prev_keys, g_current_keys);
+      g_current_keys->fill(false);
+
+      while(const uint8 key = input::poll())
+      {
+        if(key == '\0')
+          break;
+
+        g_current_keys->at(key) = true;
+      }
+    }
```

Comment:
I have no idea how yet but I would try to shield this function from the client.
Maybe the same implementation we came up with for delta time ( if this was even applied ).
You don't want the client to be able to call this, client should only be able to retrieve the key state.

Reply:
I put it in an internal namespace as we discussed